selfdrive/locationd/locationd.cc

```diff
@@ -0,0 +1,344 @@
+      }
+    }
+  }
+}
+
+void Localizer::handle_gps(double current_time, const cereal::GpsLocationData::Reader& log) {
+  // ignore the message if the fix is invalid
+  if (log.getFlags() % 2 == 0) {
+    return;
+  }
+
+  this->last_gps_fix = current_time;
+
+  Geodetic geodetic = { log.getLatitude(), log.getLongitude(), log.getAltitude() };
+  this->converter = std::make_shared<LocalCoord>(geodetic);
+
+  VectorXd ecef_pos = this->converter->ned2ecef({ 0.0, 0.0, 0.0 }).to_vector();
+  VectorXd ecef_vel = this->converter->ned2ecef({ log.getVNED()[0], log.getVNED()[1], log.getVNED()[2] }).to_vector() - ecef_pos;
+  MatrixXdr ecef_pos_R = Vector3d::Constant(std::pow(3.0 * log.getVerticalAccuracy(), 2)).asDiagonal();
+  MatrixXdr ecef_vel_R = Vector3d::Constant(std::pow(log.getSpeedAccuracy(), 2)).asDiagonal();
+
+  this->unix_timestamp_millis = log.getTimestamp();
+  double gps_est_error = (this->kf->get_x().head(3) - ecef_pos).norm();
+
+  VectorXd orientation_ecef = quat2euler(vector2quat(this->kf->get_x().segment<STATE_ECEF_ORIENTATION_LEN>(STATE_ECEF_ORIENTATION_START)));
+  VectorXd orientation_ned = ned_euler_from_ecef({ ecef_pos(0), ecef_pos(1), ecef_pos(2) }, orientation_ecef);
+  VectorXd orientation_ned_gps = Vector3d(0.0, 0.0, DEG2RAD(log.getBearingDeg()));
+  VectorXd orientation_error = (orientation_ned - orientation_ned_gps).array() - M_PI;
+  for (int i = 0; i < orientation_error.size(); i++) {
+    orientation_error(i) = std::fmod(orientation_error(i), 2.0 * M_PI);
+    if (orientation_error(i) < 0.0) {
+      orientation_error(i) += 2.0 * M_PI;
+    }
+    orientation_error(i) -= M_PI;
+  }
+  VectorXd initial_pose_ecef_quat = quat2vector(euler2quat(ecef_euler_from_ned({ ecef_pos(0), ecef_pos(1), ecef_pos(2) }, orientation_ned_gps)));
+
+  if (ecef_vel.norm() > 5.0 && orientation_error.norm() > 1.0) {
+    std::cout << "Locationd vs ubloxLocation orientation difference too large, kalman reset" << std::endl;
+    this->reset_kalman(NAN, initial_pose_ecef_quat, ecef_pos);
+    this->kf->predict_and_observe(current_time, KIND_ECEF_ORIENTATION_FROM_GPS, { initial_pose_ecef_quat });
+  } else if (gps_est_error > 50.0) {
+    std::cout << "Locationd vs ubloxLocation position difference too large, kalman reset" << std::endl;
+    this->reset_kalman(NAN, initial_pose_ecef_quat, ecef_pos);
+  }
+
+  this->kf->predict_and_observe(current_time, KIND_ECEF_POS, { ecef_pos }, { ecef_pos_R });
+  this->kf->predict_and_observe(current_time, KIND_ECEF_VEL, { ecef_vel }, { ecef_vel_R });
+}
+
+
+void Localizer::handle_car_state(double current_time, const cereal::CarState::Reader& log) {
+  this->speed_counter++;
+
+  if (this->speed_counter % SENSOR_DECIMATION == 0) {
+    this->kf->predict_and_observe(current_time, KIND_ODOMETRIC_SPEED, { (VectorXd(1) << log.getVEgo()).finished() });
+    this->car_speed = std::abs(log.getVEgo());
+    if (log.getVEgo() == 0.0) {
```

Comment:
This float comparison is fishy. Probably should be compared to something like 1e-3. The python version had the same issue, though. So maybe fix this in a separate PR.